packages/climate-ref-pmp/src/climate_ref_pmp/diagnostics/enso.py

```diff
@@ -0,0 +1,230 @@
+import json
+import os
+from collections.abc import Collection, Iterable
+from typing import Any
+
+from loguru import logger
+
+from climate_ref_core.constraints import AddSupplementaryDataset
+from climate_ref_core.datasets import DatasetCollection, FacetFilter, SourceDatasetType
+from climate_ref_core.diagnostics import (
+    CommandLineDiagnostic,
+    DataRequirement,
+    ExecutionDefinition,
+    ExecutionResult,
+)
+from climate_ref_pmp.pmp_driver import _get_resource, process_json_result
+
+
+class ENSO(CommandLineDiagnostic):
+    """
+    Calculate the ENSO performance metrics for a dataset
+    """
+
+    facets = ()
+
+    def __init__(self, metrics_collection: str, experiments: Collection[str] = ("historical",)) -> None:
+        self.name = metrics_collection
+        self.slug = metrics_collection.lower()
+        self.metrics_collection = metrics_collection
+        self.parameter_file = "pmp_param_enso.py"
+        self.obs_sources: tuple[str, ...]
+        self.model_variables: tuple[str, ...]
+
+        if metrics_collection == "ENSO_perf":
+            self.model_variables = ("pr", "ts", "tauu")
+            self.obs_sources = ("GPCP-Monthly-3-2", "TropFlux-1-0", "HadISST-1-1")
+        elif metrics_collection == "ENSO_tel":
+            self.model_variables = ("pr", "ts")
+            self.obs_sources = ("GPCP-Monthly-3-2", "TropFlux-1-0", "HadISST-1-1")
+        elif metrics_collection == "ENSO_proc":
+            self.model_variables = ("ts", "tauu", "hfls", "hfss", "rlds", "rlus", "rsds", "rsus")
```

Comment:
Do you want to run the diagnostic if you have a subset of variables available?

i.e. is the diagnostic reliant to missing data and will still run, but obviously not produce all outputs?

There is a constraint `RequireFacets` that can be used to require a set of values for a given facet.

This added to L83 would require all of the variables to be available before trying to make a run, but maybe you could still run if you only had `ts` for example.
```
RequireFacets("variable_id", self.model_variables),
```

Reply:
Good to learn about `RequireFacets`. I think it is fine to run a subset of metrics when not all variables are available, but let me double check if lack of variables could result any error, in particular when checking JSON output for CMEC. I will keep you posted for the testing.

Reply:
Confirmed that subset of metrics running okay without any issue. The package determines which diagnostics/metrics can be calculated based on given variables. The CMEC check for JSON output had no issue when subset of metrics were calculated.